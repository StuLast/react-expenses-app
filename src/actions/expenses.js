import { push, ref, get, remove, update } from "firebase/database";
import database from "../firebase/firebase";
import { ADD_EXPENSE, REMOVE_EXPENSE, EDIT_EXPENSE, SET_EXPENSES } from "./types";

//ACTIONS
//-------

//ADD_EXPENSE
const addExpense = (expense) => ({
  type: ADD_EXPENSE,
  expense,
});

const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    return push(ref(database, `users/${uid}/expenses`), expense).then((ref) => {
      dispatch(
        addExpense({
          id: ref.key,
          ...expense,
        })
      );
    });
  };
};

//REMOVE_EXPENSE

const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const dataPath = `users/${uid}/expenses/${id}`;
    const expensesRef = ref(database, dataPath);
    return remove(expensesRef).then(() => {
      dispatch(
        removeExpense({
          id,
        })
      );
    });
  };
};

const removeExpense = ({ id } = {}) => ({
  type: REMOVE_EXPENSE,
  id,
});

//EDIT_EXPENSE
const editExpense = (id, updates) => ({
  type: EDIT_EXPENSE,
  id,
  updates,
});

const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const dataPath = `users/${uid}/expenses/${id}`;
    const dataRef = ref(database, dataPath);
    return update(dataRef, updates)
      .then(() => {
        dispatch(editExpense(id, updates));
      })
      .catch((err) => { });
  };
};

//SET_EXPENSES

const startSetExpenses = () => {
  return (dispatch, getAuth) => {
    const uid = getAuth().auth.uid;
    const expensesRef = ref(database, `users/${uid}/expenses`);
    return get(expensesRef).then((snapshot) => {
      const expenseArray = [];

      snapshot.forEach((childSnapshot) => {
        expenseArray.push({
          id: childSnapshot.key,
          ...childSnapshot.val(),
        });
      });

      dispatch(setExpenses(expenseArray));
    });
  };
};

const setExpenses = (expenses) => ({
  type: SET_EXPENSES,
  expenses,
});

export {
  addExpense,
  startAddExpense,
  setExpenses,
  startSetExpenses,
  removeExpense,
  startRemoveExpense,
  editExpense,
  startEditExpense,
};
