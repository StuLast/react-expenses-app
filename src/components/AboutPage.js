import React from "react";
import { Link } from "react-router-dom";

export const AboutPage = () => <div></div>;

export default AboutPage;
