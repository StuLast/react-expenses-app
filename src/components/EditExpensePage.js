import React from "react";

const EditExpensePage = (props) => {
  console.log(props.match.params);
  return (
    <div>
      <p>This is from my Edit Expense Component</p>
      <p>Editing the expense with ID of {props.match.params.id}.</p>
    </div>
  );
};

export default EditExpensePage;
