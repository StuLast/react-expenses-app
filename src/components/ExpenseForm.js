import React from 'react';;

class ExpenseForm extends React.Component {

  state = {
    description: "",
    note: "", 
    amount: 0
  };

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    if(amount.match(/^\d*(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  };

  onSubmit = (e) => {
    e.preventDefault();
  };

  render () {
    return (
    <div>
      <form>
        <input 
          type="text"
          placeholder="Description"
          autofocus
          value = {this.state.description}
          onChange = {this.onDescriptionChange}
        />
        <input
          type="text"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <textarea
          placeholder="Add a note for your expense (optional)"
          value = {this.state.note}
          onChange = {this.onNoteChange}
        >
        </textarea>
        <button onClick={this.onSubmit}>Add Expense</button>
      </form>
    </div>
    )
  };
}

export default ExpenseForm;
