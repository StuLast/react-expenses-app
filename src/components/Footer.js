import React from "react";
import { Link } from "react-router-dom";

export const Footer = () => (
  <footer className="footer">
    <div className="container footer__top ">
      <div className="footer__column">
        <h3 className="footer__heading">Links:</h3>
        <Link className="footer__link" to="/about">
          About this demo
        </Link>
      </div>
      <div className="footer__column"></div>
      <div className="footer__column">
        <h3 className="footer__heading">Privacy:</h3>
        <Link className="footer__link" to="/privacy">
          Privacy
        </Link>
        <Link className="footer__link" to="/cookies">
          Cookies
        </Link>
      </div>
    </div>
    <div className="container footer__bottom">
      <div>
        Powered by{" "}
        <a
          className="footer__link"
          href="https://spycedconcepts.uk"
          target="_blank"
        >
          Spyced Concepts
        </a>
        . Programmed by{" "}
        <a
          className="footer__link"
          href="https://stulast.co.uk"
          target="_blank"
        >
          Stu Last
        </a>
      </div>
    </div>
  </footer>
);

export default Footer;
