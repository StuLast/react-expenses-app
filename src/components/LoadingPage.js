import React from "react";

export const LoadingPage = () => (
  <div className="loader">
    <img
      className="loader__image"
      src="/images/loader.gif"
      alt="loading screen animation"
    />
    <div>Loading...</div>
  </div>
);

export default LoadingPage;
