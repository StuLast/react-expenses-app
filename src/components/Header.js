import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";

import { startLogout } from "../actions/auth";

export const Header = ({ startLogout }) => (
  <header className="header">
    <div className="container">
      <div className="header__content">
        <Link className="header__title" to="/dashboard">
          <h1>Expensologist</h1>
        </Link>
        <button className="button button--link" onClick={startLogout}>
          Log out
        </button>
      </div>
      <div className="header__data-warning">
        This app is only a demo app. Data and data structures may be deleted or
        changed without warning.
      </div>
    </div>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout()),
});

export default connect(undefined, mapDispatchToProps)(Header);
