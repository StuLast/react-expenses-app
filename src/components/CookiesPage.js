import React from "react";
import { Link } from "react-router-dom";

import Header from "../components/Header";
import Footer from "../components/Footer";

export const CookiesPage = () => (
  <div>
    <Header />
    <div className="container">
      <h1>Cookies Page</h1>
      <h2> What are cookies?</h2>
      <p>
        Cookies are small text files that websites can store on your browser to
        help with the functionality of the site and provide additional services,
        such as shopping carts, advertising and security.
      </p>
      <h2> What cookies are used on this website? </h2>
      <p>
        When a user logs in with their Google profile, Google stores cookies on
        the browser to keep a track of authentication and data usage. Spyced
        Concepts does not collect any personal information, or store any of it's
        own cookies on this demo website.
      </p>
      <h2> Who is responsible for the cookies on this website?</h2>
      <p>
        All the cookies stores on this website are generated by Google and are
        directly related to the use of your Google account to sign in and use
        the demo services.
      </p>
      <h2> What if I don't want to use cookies.</h2>
      <p>
        To get rid of any cookies that have been stored whilst using this site,
        you can search for cookie management relating to your browser, and
        follow instructions on deleting cookies for particular websites.
      </p>
      <p>
        If you don't wish Google to store cookies when you use their login
        systems, then you can choose alternative ways to login to many sites. At
        the moment, however, this website only accepts logging in via your
        Google profile, if you have one.d
      </p>
    </div>
    <Footer />
  </div>
);

export default CookiesPage;
