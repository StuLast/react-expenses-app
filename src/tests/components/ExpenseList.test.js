import React from 'react'
import { shallow } from 'enzyme';

import { ExpenseList } from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";


describe("Render ExpenseList component", () => {
  it("should render ExpenseList with expenses", () => {
    const wrapper = shallow(<ExpenseList expenses={expenses}/>)
    expect(wrapper).toMatchSnapshot();
  });

  it("should render ExpenseList with no expenses", () => {
    const wrapper = shallow(<ExpenseList expenses={[]}/>)
    expect(wrapper).toMatchSnapshot();
  });
});

