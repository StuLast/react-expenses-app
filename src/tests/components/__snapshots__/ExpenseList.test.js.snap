// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ExpenseList component should render ExpenseList with expenses 1`] = `
<div
  className="container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <ExpenseListItem
      amount={195}
      createdAt={0}
      description="Gum"
      id="1"
      key="1"
      note="Mmmm - Chewy"
    />
    <ExpenseListItem
      amount={109500}
      createdAt={-345600000}
      description="Rent"
      id="2"
      key="2"
      note="1st Month Rent + Deposit"
    />
    <ExpenseListItem
      amount={4500}
      createdAt={345600000}
      description="Credit Card"
      id="3"
      key="3"
      note="How much???"
    />
    <ExpenseListItem
      amount={20000}
      createdAt={432000000}
      description="Hardware"
      id="4"
      key="4"
      note="Renderer Unit"
    />
    <ExpenseListItem
      amount={22000}
      createdAt={518400000}
      description="Hardware"
      id="5"
      key="5"
      note="3D printer unit"
    />
  </div>
</div>
`;

exports[`Render ExpenseList component should render ExpenseList with no expenses 1`] = `
<div
  className="container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No expenses recorded yet
      </span>
    </div>
  </div>
</div>
`;
